### PR TITLE
Credentials java

### PR DESCRIPTION
**Pull request by jenow**
Must be merged after #30 
Implements crendentials methods:
- createMyCredentials
- updateMyCredentials
- getMyCredentials
- deleteMyCredentials
- validateMyCredentials

### CODE REVIEW

File: cgo/kuzzle/create_my_credentials.go [hunk 1]
```diff
@@ -5,18 +5,25 @@ package main
...
 //export kuzzle_wrapper_create_my_credentials
 func kuzzle_wrapper_create_my_credentials(k *C.kuzzle, strategy *C.char, credentials *C.json_object, options *C.query_options) *C.json_result {
 	result := (*C.json_result)(C.calloc(1, C.sizeof_json_result))
-	opts := SetQueryOptions(options)
+	result.result = (*C._json_object)(C.calloc(1, C.sizeof__json_object))
+
+	var opts types.QueryOptions
+	if options != nil {
+		opts = SetQueryOptions(options)
+	}
```
Comment (benoitvidis) [inline cgo/kuzzle/create_my_credentials.go]: The nil check is already done in SetQueryOptions

File: templates/java/core.i [hunk 2]
```diff
@@ -29,60 +32,61 @@
...
+%extend _json_object {
+    _json_object() {
+        _json_object* j = malloc(sizeof(*j));
+        kuzzle_wrapper_json_new(&j->ptr);
         return j;
     }
 
-    ~json_object() {
+    ~_json_object() {
         free($self);
```
Comment (benoitvidis) [inline templates/java/core.i]: shouldn't we call json_object_put on ptr to free the underlying "real" json object?

**Comment by jenow** [comment]
I did encapsulated json_object because swig does not go include nested include so the json-c library is not included correctly. It was easier to encapsulate it so it would work under swig.

File: cgo/kuzzle/jsonc_wrapper.go [hunk 3]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
 
 //export kuzzle_wrapper_json_get_string
 func kuzzle_wrapper_json_get_string(jobj *C.json_object, key *C.char) *C.char {
-	var value *C.json_object
+	value := C.json_object_new_object()
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: memleak

File: cgo/kuzzle/jsonc_wrapper.go [hunk 4]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
...
 	C.json_object_object_get_ex(jobj, key, &value)
 
 	return C.json_object_get_string(value)
 }
 
 //export kuzzle_wrapper_json_get_int
 func kuzzle_wrapper_json_get_int(jobj *C.json_object, key *C.char) C.int {
-	var value *C.json_object
+	value := C.json_object_new_object()
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: memleak

File: cgo/kuzzle/jsonc_wrapper.go [hunk 5]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
...
 	C.json_object_object_get_ex(jobj, key, &value)
 
 	return C.int(C.json_object_get_int64(value))
 }
 
 //export kuzzle_wrapper_json_get_double
 func kuzzle_wrapper_json_get_double(jobj *C.json_object, key *C.char) C.double {
-	var value *C.json_object
+	value := C.json_object_new_object()
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: memleak

File: cgo/kuzzle/jsonc_wrapper.go [hunk 6]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
...
 	C.json_object_object_get_ex(jobj, key, &value)
 
 	return C.json_object_get_double(value)
 }
 
 //export kuzzle_wrapper_json_get_bool
 func kuzzle_wrapper_json_get_bool(jobj *C.json_object, key *C.char) C.json_bool {
-	var value *C.json_object
+	value := C.json_object_new_object()
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: memleak

File: cgo/kuzzle/jsonc_wrapper.go [hunk 7]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
...
 	C.json_object_object_get_ex(jobj, key, &value)
 
 	return C.json_object_get_boolean(value)
 }
 
 //export kuzzle_wrapper_json_get_json_object
-func kuzzle_wrapper_json_get_json_object(jobj *C.json_object, key *C.char) *C.json_object {
-	var value *C.json_object
+func kuzzle_wrapper_json_get_json_object(jobj *C.json_object, key *C.char) C._json_object {
+	value := C.json_object_new_object()
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: memleak

File: cgo/kuzzle/jsonc_wrapper.go [hunk 8]
```diff
@@ -105,40 +105,42 @@ func kuzzle_wrapper_json_put(jobj *C.json_object, key *C.char, content unsafe.Po
...
 	C.json_object_object_get_ex(jobj, key, &value)
 
 	return C.json_object_get_boolean(value)
 }
 
 //export kuzzle_wrapper_json_get_json_object
-func kuzzle_wrapper_json_get_json_object(jobj *C.json_object, key *C.char) *C.json_object {
-	var value *C.json_object
+func kuzzle_wrapper_json_get_json_object(jobj *C.json_object, key *C.char) C._json_object {
```
Comment (scottinet) [inline cgo/kuzzle/jsonc_wrapper.go]: See #30: I disagree with changing the returned value type

**Comment by scottinet** [comment]
I'm with @benoitvidis on this one. I'm all for adding specific SWIG specific functions and objects to make writing wrappers easier, but I do think that it should be kept apart from the C API.
It's near impossible to justify why we wrap a `json_object` pointer in an otherwise empty structure from a pure C point of view, and this makes the C developer experience considerable harder

**Comment by benoitvidis** [comment]
@jenow as you now have found a better way to handle the wrapping of json object for swig :clap:, do we still need this pr?

**Pull request closed without merging**